### PR TITLE
cpu/stm32/i2c: fix wrong speed parameters

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR is a take over from #9804 and adapted to new STM32 families added and to the different refactoring done in between.

The main idea is to not disable HSI during clock initialization when I2C is used. I think this saves a few ROM and is a bit faster during initialization compared to the initial approach where this is done in the periph initialization (so enable hsi -> disable hsi -> reenable hsi, whereas with this PR: enable hsi once).

Maybe other families that depends on this version of the I2C driver (f0, f3, l0 and l1) could be adapted but I didn't check yet.

### Testing procedure

- Green CI
- I2C is working (probably better)

### Issues/PRs references

Take over from #9804, so closes #9804

### CODE REVIEW

**Comment by fjmolinas** [comment]
@MrKevinWeiss can the HIL be ran here?

**Comment by MrKevinWeiss** [comment]
It can, but I don't have clock checking for the i2c tests yet so I don't know how useful it would be.

**Comment by MrKevinWeiss** [comment]
Maybe something for @MarcelStenzel to add?

**Comment by benpicco** [comment]
Murdock found some issues

**Comment by benpicco** [comment]
@MrKevinWeiss I think it would suffice to verify that I2C still works at all with this change :wink:

**Comment by MrKevinWeiss** [comment]
Maybe [this](https://hil.riot-os.org/results/robot_tests/187/overview) is better... Compared to [nightlies](https://hil.riot-os.org/results/nightly/latest/overview) I don't see much difference at least for the nucleos.

**Comment by fjmolinas** [comment]
> Maybe [this](https://hil.riot-os.org/results/robot_tests/187/overview) is better... Compared to [nightlies](https://hil.riot-os.org/results/nightly/latest/overview) I don't see much difference at least for the nucleos.

Does that mean master is broken?

**Comment by fjmolinas** [comment]
I don't have many `stm32` boards at hand, any chance you could give a few a test @aabadie?

**Comment by aabadie** [comment]
Before someone ACKs this PR, I'd like to mention that the proposed approach has a power consumption penalty: on STM32 boards that provides an HSE, when using I2C, the HSI will always be running when not sleeping while in master it's shutdown.

IMHO, the ideal solution would be to compute these parameters once (there's a formula in the manual, but well) based on the bus clock frequency. That will increase RAM/ROM usage but it will guarantee that HSI is not running when not used as sysclk source (either directly or via PLL).

**Comment by aabadie** [comment]
Tested on G4, L4 and F7. They are working (I had to push a fix for F7 though).

**Comment by aabadie** [comment]
Now it would be good to test on G0, L5, WB and WL, to be complete.

**Comment by aabadie** [comment]
Tested on G0, L5, WB and WL: reading I2C sensor values works like a charm.

**Review by fjmolinas — APPROVED** [review]
ACK, trusting @aabadie testing.

**Comment by benpicco** [comment]
@aabadie I think you are the best to make the trade-off wrt the HSI question here. So previously HSI was only running when the bus was acquired, now it's always running?
How much does this affect the power draw? And why can't we only enable HSI on demand anymore with this?

**Comment by aabadie** [comment]
> So previously HSI was only running when the bus was acquired, now it's always running?

No, before the I2C was simply clocked by SYSCLK which is by default not corresponding to the configured I2C parameters.

> How much does this affect the power draw? And why can't we only enable HSI on demand anymore with this?

I can try to monitor power consumption but I'm not an expert here.
We could enable HSI on demand, but I fear this would a little performance penalty when acquiring/releasing the bus.

**Comment by aabadie** [comment]
Hmm, and this needs a rebase

**Comment by aabadie** [comment]
Rebased this PR and moved the HSI enable/disable strategy to the acquire/release function of the I2C peripheral driver. I made quick tests on L4 and L0 (HSI always enabled on the L0 board I tried) and the behavior is correct: HSI is enabled if not already enabled and only disabled if the HSI was off before acquiring the bus. Since an I2C bus communication is synchronous, using a static variable for the state should be safe but I need someone else's opinion on this point.

I think some changes are still missing in this PR: I2C configs must be adapted on more boards but I think Murdock will catch them.

**Comment by fjmolinas** [comment]
@MrKevinWeiss this seems to be the case for some HIL in the loop testing (only for stm32 BOARDS), could you run some tests?

**Comment by MrKevinWeiss** [comment]
> @MrKevinWeiss this seems to be the case for some HIL in the loop testing (only for stm32 BOARDS), could you run some tests?

Sure thing, [here is the current run](https://hil.riot-os.org/jenkins/blue/organizations/jenkins/robot_tests/detail/robot_tests/189/pipeline).  I am am only running the i2c tests from robot fw, the PR feedback will view it as a single test (even though there are many) and if there is a failure it will say the whole thing failed...  I will post a link to easier to read test results once complete.

**Comment by MrKevinWeiss** [comment]
[Here are the results](https://hil.riot-os.org/results/robot_tests/189/overview).

So far it looks similar to master, the only difference looks like the `nucleo-l432kc` is locked up, I will give it a retry since it works on master.

**Comment by MrKevinWeiss** [comment]
The rerun of the `nucleo-l432kc` works, it seems it was an infra problem.  No difference between master and this PR for the robot tests.

We don't yet have a clock speed metrics for the i2c, maybe @MarcelStenzel would like to add it to his list :wink:

**Comment by fjmolinas** [comment]
> Rebased this PR and moved the HSI enable/disable strategy to the acquire/release function of the I2C peripheral driver. I made quick tests on L4 and L0 (HSI always enabled on the L0 board I tried) and the behavior is correct: HSI is enabled if not already enabled and only disabled if the HSI was off before acquiring the bus. Since an I2C bus communication is synchronous, using a static variable for the state should be safe but I need someone else's opinion on this point.
> 
> I think some changes are still missing in this PR: I2C configs must be adapted on more boards but I think Murdock will catch them.

I think thats fine:

- stm32 is only 32 bit, so every read/write should be atomic
- they are done after the mutex, so there can't me multiple accesorss
- AFAIU the mutex call results in a memory barrier

**Comment by fjmolinas** [comment]
@aabadie if murdock is going to catch all missing adaptions please squash once thats done

**Review by fjmolinas — APPROVED** [review]
ACK, HIL testing provided, @aabadie did some testing as well.

File: cpu/stm32/include/cpu_conf_stm32_common.h [hunk 1]
```diff
@@ -66,6 +62,31 @@ static const i2c_timing_param_t timing_params[] = {
         .sdadel = 0x0,      /* t_SDADEL = 0ns   */
         .scldel = 0x1,      /* t_SCLDEL = 250ns */
     }
+#elif defined(CPU_FAM_STM32L4) || defined(CPU_FAM_STM32F7) || \
```
Comment (aabadie) [inline cpu/stm32/include/cpu_conf_stm32_common.h:65]: Open question: should I also include f0 and f3 in this list ? And then drop f0 and f3 specific parameters. I haven't tried I2C on these platforms with this PR and with these parameters but I think that's worth.

Comment (fjmolinas) [reply cpu/stm32/include/cpu_conf_stm32_common.h:65]: Please do

Comment (aabadie) [reply cpu/stm32/include/cpu_conf_stm32_common.h]: Done, with some extra cleanups: the `cpu_conf_common.h` was not needed anymore. I moved its content in `cpu_i2c.h`, guarded by the same CPU family check and removed the file.
Let's see if that passes Murdock.

**Comment by fjmolinas** [comment]
ACK still stands

**Comment by aabadie** [comment]
> please squash

Done

**Pull request merged**